### ISSUE TITLE
firewall: global prefix in rules needs to be changed manually when IPv6 upstream prefix changes

### ISSUE THREAD

**Issue by RyuunoAelia** [opnsense/core#2544]
While playing a bit with NPTv6 I found a a workflow issue. The main and only reason to use NPTv6 is to avoid changing your local configuration (firewall rules, etc) when your upstream changes.

However, as things are done in opnsense right now, you need to manually update the NPTv6 rule each time the upstream changes.

This is fine for cases where you have complete control over when the prefix change occurs (multi-homed setup, etc). However, when the prefix may change without your knowledge (say a 6rd setup, or ISP providing ipv6 connectivity with short lived DHCP leases), having to manually update the rule painful.

I understand selecting the IPv6 might be problematic to implement in general. An interface may have multiple IPv6 at a point in time (even if you ignore link-local addresses). So maybe having a way to select on an interface if this interface will have dynamic NPTv6 rules might be a more straightforward flow for setting up things. But since I do not know much about the internals of the UI, I cannot give a pertinent insight.

My full context is described in #2538.

**Comment by fichtner** [comment]
This is also useful for DHCPv6 and SLAAC, but will require more tinkering.

**Comment by fichtner** [comment]
Not for the rules currently. It doesn't allow tracking the interface prefix change.

**Comment by marjohn56** [comment]
So maybe we try splitting the rules into prefix/suffix..,, Alias for Prefix+suffix. It does make it complex for the user too, but that's the problem with a changing prefix, it's not for the faint-hearted.

**Comment by fichtner** [comment]
We already have a larger change in that area to make 6RD usable for "inet6" type rules and need to see if that is alright before we stack another layer on top, but generally an automatic prefix adaption option should be possible

**Comment by chris42** [comment]
Having similar troubles discussed with marjohn56 in https://forum.opnsense.org/index.php?topic=9438.0

Being completely obvious on how technically the DHCPv6, SLAAC, RADV and the firewall works: Since the clients in the network negotiate their IPv6 couldn't Opnsense create a central register in the DNS, that is build during the negotiation process? This would keep the changing IP information centralized and up to date in realtime. (This would cover prefix changes as well as changes by e.g. privacy extensions)
Then you could use that register with Hostname in the firewall rules?

Basically using the Alias, but not having a 5min delay between resolving the hostname, but accessing the register (which should be up-to-date) of IPv6s in the network?

**Comment by fichtner** [comment]
Since we need to force a rules update on IP changes it's probably better to read the current IPv6 and merge the prefix accordingly and reload the rules. That should cover 99% of the use cases. The only tricky thing is integrating this logic and how to present it to the user (if applicable).

**Comment by chris42** [comment]
I would put an option next to the prefix delegation option to enable automatic prefix update to firewall rules. Then have in the Firewall rule next to the IP window an option to use delegated prefix. IP window then only holds the suffix part.
Then you could control if you actually want to use that feature next to the prefix delegation and then control per firewall rule, if it is a rule, that uses this feature.

**Comment by RyuunoAelia** [comment]
To answer the "use hostname in rules" remark. It is not recommended by pf (the firwall software) documentation (because in general the firewall is started before any other services thus there is no DNS available to query).

The thing is, we don't need a convoluted thing like DNS, since opnsense (and pfsense for that matter) already regenerates all firewall rules when the IPv4 changes via DHCP, so we only need to have a way to map the IPv6 change too. With IPv6 it makes sense to split the network part and the machine part of addresses I think.
Event if both parts of IPv6 addresses can be dynamic since the internal network controls the machine part we can forget about it, and only concerntrate on managing changes in the network part (A.K.A. prefix).

The way I see things, you don't even need to change that much to UI, just make the IPv6 firewall rule editing look more like the IPv4 rule, with the drop-down list for Type of Source/Destination with a "Interface subnet" option and an optional "machine part of address" field or something like that. Then there's a bit of a tricky mangling of the input to regenerate the full address to pass to the pf command to do in the *.inc files.

**Comment by staeglis** [comment]
What's the current state of this issue? Will this be really solved in 19.7?

**Comment by cpw** [comment]
Is there anything I can do to help and or test anything?

**Comment by chris42** [comment]
This is quite disappointing. Together with the other IPv6 changes I was really looking forward to the 19.7
Now it is another 6 month.

**Comment by fichtner** [comment]
I'm happy to say to be disappointed by disappointment for lack of hands on deck on issues that are disappointing to be left disappointed in the first place. If nobody cares enough the community will go nowhere and disappointment is really displaced because there is a general lack of drive.

**Comment by chris42** [comment]
Than that is something that should be made public, that developers, funds, etc. are needed.

For me as a user, I cannot tell. I can only see the features that are rolled out and what is pushed back. IPv6 sadly/disappointingly the latter.

**Comment by fichtner** [comment]
No, as a user you can do better than simply share your subjective disappointment, especially on a major release day where a couple of things have been accomplished despite said disappointment.

**Comment by chris42** [comment]
I agree that the timing could have been better tomorrow. 
Just see that I happily waited and tinkered with my setup for over a year now to get a proper IPv6 support. Something I cannot solve on my own, but need a developer to do. Until I saw your release E-Mail, I only knew it was tagged for 19.7 for half a year and came here to see, that it was moved 2 weeks ago.

So if you ask me to do better in communication, I will happily try to. But then it follows, that I ask you to do better in it as well and look into proper expectation management and tell us what Opnsense would need as support.

So I apologize for throwing my unfiltered disappointment out there. Please see it, as how much I like Opnsense and how much I want it to succeed. (there is no irony or sarcasm here)

**Comment by fichtner** [comment]
Thanks for the explanation. Please consider me responding immediately to questions as "expectation management". I don't know what "proper" is, it looks to be in the eye of the beholder. For myself "proper" would mean transparent and precise.

The problem we see is we have about 150 open tickets at any point in time. Some of them are support questions, setup issues and others are bugs and features. For 19.7 we tackled 112 of bugs and features alone. Many if not all of them are more important than this, probably due to a mix of complexity, previous state and underlying code base that does not recognise the shifting prefix fact that the ISP for all intents and purposes only does to annoy customers as there is no real need for it.

There is also private life and these things naturally do not help with having contributors contribute more.

Lastly, we have this bug tracker not as a sort of vending machine where anyone can press a button and a feature drops out in the next version. We use tagging to make clear priorities. Tagging a milestone is good. It doesn't necessarily mean work will be finished in time. Some features took years to pull off. All we really have is time and I appreciate all suspending of judgement on these subjectively pressing matters.


Cheers,
Franco

**Comment by staeglis** [comment]
Is there still a realistic chance to see this feature in 20.1?

**Comment by fichtner** [comment]
Planning is provided as "best effort" for the next upcoming release. In this particular case, I don't think this will make 20.1.

**Comment by chris42** [comment]
To maybe prevent the same issue as with 19.7: If this is not making 20.1, the milestone should be removed, just for transparency and precision.

**Comment by BigSnicker** [comment]
Don't have the expertise to help, but I donated some cash and would love to see this feature to help out vs. my insane prefix-changing, 10m subscribers, largest-in-Canada ISP (Rogers).

Found this thread from https://forum.opnsense.org/index.php?topic=11011.0

**Comment by fichtner** [comment]
@chris42 my best contribution here for the moment is to keep this in the milestone queue otherwise the ticket will be dropped eventually if nobody works on it. the only thing that helps more is code...

**Comment by marjohn56** [comment]
@fichtner - Let's try and work out something for this, and other related issues to do with dynamic IPv6, we've got quite a long way towards resolving this is and related issues with the work we did early last year with the static dhcpd6 changes.

It would seem logical  to have a single function somewhere that gets called and updates the fw, unbound and others when the prefix changes; possibly a unified function that that gets called whether it's static or dynamic and does the work of all the different calls that are presently in place scattered about other files. I should have some time to play with this again and come up with ideas you can pull to pieces until we find one that fits.

**Comment by AdSchellevis** [comment]
@marjohn56 sorry for stepping in, but would it be possible to construct a test setup simulating the issue or issues (two OPNsense xml's and steps to reproduce)? 

I'm getting the feeling we're trying to solve multiple (possibly related) issues at once, which has a high risk of ending up in patchwork. There likely is some sort of rfc beneath a lot of this, it would be good to stay as close as possible to how things are originally intended to work (unwinding these things later has the habit of ending up in a lot of work and new user challenges)

**Comment by marjohn56** [comment]
Don't be sorry, more on deck the better, many hands make light work ( or too many chefs spoil the broth! ).  Many of the issues are related, and to do with manually assigned dhcpd6 with a tracking interface. It was a feature a lot of users asked for and we got that bit working early in 19.7.1, what we missed were the associated requirements to change a few things in Unbound etc when we have a change, static mapping is one that comes to mind. It's not my intention to try and get a rapid fix out for 20.2, but to look at that area as a whole and see if we can push it forward in a way that we have a single inc file that contains the glue that brings together the calls to the various other functions needed when an interface prefix changes. That way it should be easier to maintain going forward. 

The error I/we made before was not realising the chain reaction that would happen in the likes of unbound etc. The firewall area is a baby you can bounce on your lap. :) 

What we need first is a framework of what needs to be included/updated when a prefix changes.

**Comment by Vagor96** [comment]
Would it be possible to merge the fix described in #3657 as a workaround? The DNS entries could then be used in firewall rules until a better system is in place. At the moment the firewall is really barely usable for IPv6 on connections with dynamic prefix.

**Comment by HikariWS** [comment]
I had described my situation on https://forum.opnsense.org/index.php?topic=19405.0 and believe that this would help me a lot!

tnx anybody who's making this happen :)

**Comment by spi43984** [comment]
Not sure I understand that right.

This ticket #2544 is about the feature to use dynamic prefixes in rule, e. g. use some variable for the current prefix in rules.
That feature was planned for 21.1.

The ticket #4642 is about the missing prefix in DNS replies.

If #4642 was fixed DNS queries would work. I could use DNS names in firewall rules then but the issue from this ticket #2544 still would exist and there wouldn't be any way to use firewall rules with IPv6 addresses and dynamic prefixes.

**Comment by fichtner** [comment]
#2544 It was planned for the next milestone we had... we only have a milestone for the next version. We can't tackle all the things on all the time on core time and so now we have a "community" milestone to better reflect the lack of outside contributions regarding the request.

#4642 yes

> If #4642 was fixed DNS queries would work. I could use DNS names in firewall rules then but the issue from this ticket #2544 still would exist and there wouldn't be any way to use firewall rules with IPv6 addresses and dynamic prefixes.

Not really. The prefix change still breaks a setup easily so you have the DNS leases on the old prefix even when fixed to use the prefix when it was last configured.... In the end, the shifting prefix is an annoyance that will not go away easily even if a lot of work is put into dodging its effects from our end.

**Comment by spi43984** [comment]
Ok, thanks. Understood. Dynamic prefixes are an annoyance.

Is there a workstream (or some ticket) though looking into how to technically best support dynamic prefixes? I don't expect all Internet providers to switch to static prefixes for all their customers.

**Comment by froogl** [comment]
If I understand this correctly the actual problem is that, contrary to other firewall solutions out there, FreeBSD's packet filter (pf) does neither support suffix matches nor outgoing interface matches. What we'd actually need is support for firewall rules like in Linux, just like this (nftables example):

```
nft add rule inet filter iifname $WAN oifname $LAN ip6 daddr & ::ffff:ffff:ffff:ffff == ::f2e5:3a81:15cf:8e16 tcp dport ssh accept
```

That rule completely ignores the prefix and only matches based on interface names and IPv6 tokens/suffixes. I think FreeBSD's ipfw2 at least has support for matching the outgoing interfaces, however I am not sure about IP suffix matches. In the end this is a limitation of the underlying operating system (BSD) which is showing it's age in this case, as it's lacking support for features required by modern internet protocols and their implementation in the real world.

I don't think this issue can be easily resolved at the OPNsense project level. It could only be worked around by a hell of shell scripts and triggers (which the prefix handling kind of is already) but a clean solution will probably not be achievable with the current state of the FreeBSD networking stack.

So in the end there are a couple of options:

1. Go the scripting way (e.g. reloading firewall rules on prefix changes). This is a quick, easy and hard to maintain solution. And it's ugly as hell, as it's just working around the core problem.
2. Invest a lot of time in improving `pf`, so it actually supports firewall rules with suffix matches and interface matches. If I understand the architecture of `pf` correctly, especially the interface matching part is out of scope of the software. If I am correct with this one this would definitely be a deal-breaker.
3. Switch to one of the other firewalling frameworks which FreeBSD offers. Not sure if there is any which actually supports the required features.
4. Switch to a completely new OS base, such as Linux. This way we could get a more advanced and modern networking and firewall stack. A neat side-effect would also be much better driver support in general and better throughput for many drivers.

**Comment by spi43984** [comment]
I'm not much into BSD's packet filter so I can't tell if this might be an approach or not: with pfctl it seems to be possible to rewrite the prefix after it changes: https://forums.freebsd.org/threads/how-to-apply-non-numeric-mask-to-ipv6-address.69829/

Dynamic prefixes are a pain but will accompany us for quite some time, see https://tools.ietf.org/html/draft-ietf-v6ops-slaac-renum-04#ref-GERMAN-DP and especially in countries like Germany for data privacy reasons (in German language only): http://www.bfdi.bund.de/SharedDocs/Publikationen/Entschliessungssammlung/DSBundLaender/84DSK_EinfuehrungIPv6.pdf?__blob=publicationFile

The document by state's and countries' data privacy officers suggests the Internet providers implement at least dynamic prefixes for data privacy reasons. In case they offered fixed prefixes they also need to offer other ways to get the IPv6 changed by the user. So the easiest impementation is just to offer a dynamic prefix. Too bad.

As I move away from pfsense to something else I stumbled across the dynamic IPv6 issue in opnsense. As I understand that is not easy to fix in opnsense, I wonder how other's have solved this issue? One of course could be to use native IPv6 only for outgoing client connections and implement an additional IPv6 tunnel broker with a fixed prefix for server connectivity from the Internet. But maybe there are smarter ways?

**Comment by HikariWS** [comment]
I didn't know FreeBSD has this limitation, very sad. This is offtopic, but what Linux distro other than OpenWRT is best equivalent to OPNsense in features?

On OpenWRT, prefix and suffix are handled separately. Global prefix comes from ISP and "just works", and ALAIK we can't manually set the remaining bits when it's shorter than 64bit. The suffix is set on /etc/conf/dhcp on each host (hostid) then DHCPv6, SLAAC and odhcp use it to build the IPv6 addr together with the prefix.

It supports stuff like "$(uci get network.globals.ula_prefix | sed -e "s/^./d/")" according to https://openwrt.org/docs/guide-user/network/ipv6/ipv6.nat6

IPv6 has this big design flaw where LAN addresses may use ISP provided global prefix. This allows ISP to mess with our private LAN, as when they provide only /64 prefix and blocks us from creating VLANs. ISPs won't stop these practices, so IMHO we'll end up needing NPTv6 so we can isolate LAN "GUA" from ISP global prefix.

**Comment by froogl** [comment]
> This is offtopic, but what Linux distro other than OpenWRT is best equivalent to OPNsense in features?

Due to FreeBSD's IPv6 shortcomings I have migrated from OPNsense to Debian with nftables, systemd-networkd, dnsmasq and wireguard. It's a very clean configuration without any hacky shell scripts. I don't need a web UI, so I am basically unlimited in features and possibilities. And I can finally saturate my 1G fibre connection, OPNsense capped the throughput at ~600 Mbits. So if you feel comfortable with the command line, that would be the way to go (replace "Debian" with the distro of your choice).

> IPv6 has this big design flaw where LAN addresses may use ISP provided global prefix.

This is not a design flaw. I still think it's great to have global addresses in my private network. As long as you have a firewall in between there is nothing to worry about :) No need to set up NPTv6 or any other mapping hacks, as long as the underlying OS/firewall handles prefix changes correctly.

**Comment by marjohn56** [comment]
The flaw  is with the ISPs, there is absolutely no reason to use changing prefixes and for ISPs it's a cash cow. If you want statics you need to pay extra for it. My suggestion, change to a decent ISP if you can. My ISP gives me a static /48 at no extra cost.

**Comment by froogl** [comment]
I personally prefer changing IPv4 addresses and IPv6 prefixes, as I consider it a privacy feature. Sometimes I actually need a new IP address/prefix to bypass IP restrictions. I also don't consider it a flaw, it's just an implementation of RFC3633/8415:

> Each prefix has an associated valid and preferred lifetime, which constitutes an agreement about the length of time over which the requesting router is allowed to use the prefix.

So if a firewall cannot properly deal with prefix lifetime and rollover, as specified by the mentioned RFCs, I'd rather consider this a flaw in the firewall implementation instead of the ISP. No offence here, I know this is definitely not trivial to implement. But my feeling is that the only thing that has been done on this matter in the past was throwing mud at the ISPs. It's clearly a software issue.

Is a static prefix easier to handle? _Sure it is_. But that does not mean that these changing prefixes are an evil thing we need to fight. And even if it was, we need to face the reality: The vast majority of ISPs in the world offer dynamic prefixes to their consumer-grade customers. And as mentioned earlier this reality is perfectly fine even from a standards perspective.

**Comment by marjohn56** [comment]
Then get on and fix it.

**Comment by fichtner** [comment]
I'm not sure how great Linux compares with threads like this https://forum.openwrt.org/t/delegated-ipv6-prefix-not-updated/56135/9

I spend quite some time to research this and you hit a lot of dead ends with pf, SLAAAC being handled in the kernel only so you can't listen for events, DHCPv6 not notifying when forcing a prefix reload on an upstream router, DHCPv6 not being able to probe for new prefix because the socket is blocked by the running DHCPv6 client already.

**Comment by froogl** [comment]
> Then get on and fix it.
 
Well, I have personally already fixed the situation for myself by migrating to a nftables-based system. And I was trying to give some options to discuss https://github.com/opnsense/core/issues/2544#issuecomment-769811809 as a starting point. I still don't know which route should be taken.

> It's not correct to say no work has been done on this, a lot of work has been done on this, but as soon as we fix one thing another pops up.

Sorry about this, I was not trying to say that no work has been done. But all I can read on the forums and Github are statements like "dynamic prefixes are [broken|wrong|evil|whatnot]_". And I wanted to clarify that they are not. They are perfectly fine if you are using the proper tools.

**Comment by spi43984** [comment]
I am coming from pfsense and have just recently migrated to opnsense. I like many features and the ease of useability.
But as I am customer of a large German ISP I am also stuck with changing prefixes. There is no way to change the ISP or use a business plan due to other limitations like lack of IPTV then. So I'll be stuck with dynamic prefixes until the ISP changes that if ever.
Feeling like standing at a crossroad I was also thinking of migrating to Debian. Another option might be (which I haven't thought through yet) is use the ISP's IPv6 for outgoing client connections (with some additional benefit of privacy) and use an IPv6 broker for incoming IPv6 connections aimed at servers.

**Comment by HikariWS** [comment]
> > This is offtopic, but what Linux distro other than OpenWRT is best equivalent to OPNsense in features?
> 
> Due to FreeBSD's IPv6 shortcomings I have migrated from OPNsense to Debian with nftables, systemd-networkd, dnsmasq and wireguard. It's a very clean configuration without any hacky shell scripts. I don't need a web UI, so I am basically unlimited in features and possibilities. And I can finally saturate my 1G fibre connection, OPNsense capped the throughput at ~600 Mbits. So if you feel comfortable with the command line, that would be the way to go (replace "Debian" with the distro of your choice).

Interesting. So Linux is indeed better than FreeBSD for IPv6?

I was searching articles and OPNsense is the top recommended FOSS OS for router, the only one using Linux seems to be IPFire. Of course I'd like a distro that's prepared for the job, but I'd not skip a lean distro either. I like WebUI for monitoring stuff, but for configuring I prefer plain txt files which I can handle on Subversion. On my current OpenWRT, I rarely use Luci for configs. But I'd need to learn a big lot to manage everything from a lean distro, which is troubling when we're talking about router.

> > IPv6 has this big design flaw where LAN addresses may use ISP provided global prefix.
> 
> This is not a design flaw. I still think it's great to have global addresses in my private network. As long as you have a firewall in between there is nothing to worry about :) No need to set up NPTv6 or any other mapping hacks, as long as the underlying OS/firewall handles prefix changes correctly.

As I said, the issue is ISP being able to dictate our LAN addresses: change prefix and breaking internal connectivity, break VLANs, etc etc. I don't want NAT as a security feature, I want some feature to decouple LAN addresses from ISP prefixes, and better multi-homing. Without workarounds.

> The flaw is with the ISPs, there is absolutely no reason to use changing prefixes and for ISPs it's a cash cow. If you want statics you need to pay extra for it. My suggestion, change to a decent ISP if you can. My ISP gives me a static /48 at no extra cost.

The flaw is the protocol allowing ISPs to abuse it. To force LAN addresses to be coupled with prefix provided by ISPs. It's useless that the protocol designer kindly asks ISPs to properly follow the protocol standard, much less to not abuse it. It's like asking them to not snuff on HTTP and DNS messages, that's why we have HTTPS and DoH. Hoping for proper ISP competition also can't be on the protocol design, it just helps ISPs under monopoly to be even more abusive.

The standard clearly "suggests" /56 prefix, but nothing stops any ISP from providing /64, and even charge extra fee for bigger one.

What we need is proper protocols/tools that gracefully decouple ISP prefix from LAN topology while inceting devices OS to prefer IPv6.

**Comment by chris42** [comment]
@HikariWS 
I think you are messing up IPv4 concepts with IPv6. IPv6 was designed with IPv6 global route able addresses for every device in mind. Also heavily using dynamic allocation and reliance on name resolution. Most services nowadays come from the IPv4 world and have troubles to adapt to that IPv6 world.
What you are looking for with you own LAN is fd00::/8. No ISP is messing with that and you are free to build a IPv4 topology with IPv6 addresses. By standard, an ISP does not have to give you a prefix at all and could just handle single IPv6s per device. Please stop mixing different concepts and creating a false narrative of ISPs that try to abuse you.

**Comment by BigSnicker** [comment]
The flaw is with the ISPs, there is absolutely no reason to use changing
prefixes and for ISPs it's a cash cow. If you want statics you need to pay
extra for it.

That's true in North America, but I saw someone write that dynamic prefixes
are becoming mandatory in the EU, driven by Germany's strict data privacy
regulations.

I wasn't able to verify that that's true (I'm sure our Dutch friends will
know), but it does seem to come up occasionally:

https://www.huntonprivacyblog.com/2012/11/16/german-dpas-adopt-resolutions-on-eu-data-protection-and-ipv6/

https://forum.openwrt.org/t/ipv6-prefix-privacy/32179

https://svs.informatik.uni-hamburg.de/publications/2012/2012-10-09IPv6-Prefix-Alteration.pdf

**Comment by spi43984** [comment]
> https://www.huntonprivacyblog.com/2012/11/16/german-dpas-adopt-resolutions-on-eu-data-protection-and-ipv6/

That's the one in proper English I was referring to earlier. German ISPs might offer fixed prefixes to end users but in that case they need to offer end users some way to easily change their static IPv6 prefixes. So thinking about some highly-sophisticated backend service to allow end users changing their IPv6 prefix, it's much easier for an ISP to allow dynamic prefixes only. I don't expect that to change soon.

**Comment by HikariWS** [comment]
@chris42 as I said this is offtopic here. I disagree on your statements, call me in pvt if you are interested on talking about it :)

> The flaw is with the ISPs, there is absolutely no reason to use changing prefixes and for ISPs it's a cash cow. If you want statics you need to pay extra for it.

I don't disagree on ISP's bad faith, or on proper needs for prefix changes as pointed by @spi43984. The issue is that IPv6 provides possibility for ISP to change prefix and/or provide /64 prefix, regardless if they are abusing the protocol or due to legal demands.

The protocol design can't allow ISP to prohibit us to use multiple VLANs or drop our connections at will, and then just hope or beg for ISPs to not do it. Some will always do, so the protocol must not support it or provide features to circumvent it.

The topic here isn't either ISPs behavior or how IPv6 is good or bad or how it could be better. Given the circumstances on the table, we must find solutions on network tools (OPNsense here).

We can't stop ISPs from abusing and we have acceptable use cases for prefix changing and even reference on the standard regarding prefix delegation lifetime, so only viable solution I see is for tools to gracefully identify when PD had changed and propagate it to all other router tools and devices under it. I know how hard it is to implement that, which is very sad, and I thank all developers who had been voluntarily working on it.

**Comment by bu7cher** [comment]
AFAIR, I've added `ext_if` option for NPTv6 module exactly for such purpose - detect changing of external prefix and reconfigure the NTPv6 instance. Does it not fit for your tasks?
  https://cgit.freebsd.org/src/commit/sys/netpfil/ipfw/nptv6/nptv6.c?id=b2b56606889cd11b155472009a991d458ff119f7

**Comment by marjohn56** [comment]
> AFAIR, I've added `ext_if` option for NPTv6 module exactly for such purpose - detect changing of external prefix and reconfigure the NTPv6 instance. Does it not fit for your tasks?
> https://cgit.freebsd.org/src/commit/sys/netpfil/ipfw/nptv6/nptv6.c?id=b2b56606889cd11b155472009a991d458ff119f7

Nice..

**Comment by maurice-w** [comment]
Nice indeed, but this once again brings us back to the issue that OPNsense doesn't use ipfw (except for traffic shaper and captive portal I think). Correct?

Same issue with NAT64: Natively supported in ipfw, but we have to use Tayga on OPNsense.

Maybe, just maybe, eventually switching to ipfw would make OPNsense more future proof? But I know it's a very complex task.

**Comment by HikariWS** [comment]
> AFAIR, I've added `ext_if` option for NPTv6 module exactly for such purpose - detect changing of external prefix and reconfigure the NTPv6 instance. Does it not fit for your tasks?
> https://cgit.freebsd.org/src/commit/sys/netpfil/ipfw/nptv6/nptv6.c?id=b2b56606889cd11b155472009a991d458ff119f7

Sorry for my noob question on github. Does this mean that OPNsense supports dynamic prefix changing under NPTv6? And what about multi-homing / multi-WAN, is it supported?

If so, could somebody make a tutorial teaching how to enable dynamic prefix support on NPTv6 with 2 WANs? I'm still using OpenWRT and only waiting OPNsense to support dynamic prefix to build a new PC for router.

**Comment by maurice-w** [comment]
No, this only means that there is a feature in FreeBSD which supports it, but OPNsense doesn't make use of this (yet).

NPT with multi-WAN works, but only with static prefixes.

**Comment by HikariWS** [comment]
Well at least we have NPTv6 working. Let's hope some day OPNsense works with multi-homing and dynamic prefixes.

**Comment by imp1sh** [comment]
How did pfSense solve that in 2.5? I heard they implemented a solution for changing prefixes. The license should allow to borrow it, right?

**Comment by froogl** [comment]
Good news people. The current dnsmasq-git tree has support for dynamic AAAA records, which means it can automatically prepend the current IPv6 prefix to statically configured addresses. :tada: That's how you use them:

```
--dynamic-host=<name>,[IPv4-address],[IPv6-address],<interface>
```

As soon as an updated dnsmasq has been released we can reference our IPv6 hosts in firewall rules, even in scenarios with dynamic prefixes. So assuming you have a server with a global (dynamic) IPv6 address sitting behind your OPNsense box and you need to add a firewall rule for that host, that's what you'd need to do:

- Use dnsmasq as your DNS server (as soon as the next stable release has been published). OPNsense already supports this.
- Add a dynamic DNS record in dnsmasq for each host you need a firewall rule for by using the new `dynamic-host` option in dnsmasq. It's up to you whether you are using a SLAAC address, a static interface token configured on the target system or an address you are handing out through DHCPv6.
- Configure a firewall alias for each host containing it's DNS name. The firewall will regularly resolve that DNS name to an IPv6 address. So your firewall rules are updated automagically on prefix changes.

**Comment by fichtner** [comment]
For reference the pfSense proposal is https://github.com/pfsense/pfsense/commit/7c4b3d3c which deals with none of the drawbacks which is basically what we do for DHCP static leases too. None of the drawbacks are being addressed and shifting prefix might break a running configuration in complex rule setups / port forwards (think about negating that address and then mismatching on it).

I suspect work needs to be done in dhcp6c to deal with seamless reloads by some means of system notification. SLAAC is also problematic because the kernel managed this all by itself originally aimed at client configuration, not routers.

**Comment by marjohn56** [comment]
@fichtner - we probably already get that marker from dhcp6c from the ENV vars. If the ENV var is REQUEST then it's a new lease / possible new address, if it's RENEW then it should still be the old lease. We also have the address/prefix pulled from the ENV var too, easy enough to compare existing with new/renewed values.

**Comment by nanosonde** [comment]
This is how somebody solved basically the same issue on RouterOS with scripting:
https://forum.mikrotik.com/viewtopic.php?t=168470
Script: https://n377.de/mikrotik-routeros-script-update-the-firewall-with-dynamic-ipv6-prefix.html

Is the OPNsense API currently "powerful" enough to achieve the same by a similar script running on the OPNsense instance?

**Comment by fichtner** [comment]
Basically what we need is a new alias type that can expand addresses according to its attached interface. The interval logic is already there and it works fine for DNS or MAC updates already.

This way we avoid hard coding addresses in rules.debug and have a mitigation strategy for prefix change detection by simply running the update at a sane but short interval.

I don’t have a lot of extra time for 21.7 development due to other priorities (installer migration) but I can do a pitch for this internally next week to see how much work it is and thus what time frame it can fit in if someone is willing to allocate the time this year.


Cheers,
Franco

**Comment by marjohn56** [comment]
I expect @maurice-w and I could play with this. So the alias would just pick up the assigned interface prefix and leave the suffixes intact... 
I've never looked at the way the aliases work, so hopefully @maurice-w has a clue. :)

**Comment by fichtner** [comment]
Yes, in the alias you put one or more suffixes and select the interface to expand it to.

For management view it is more work, but the results are far better than just merging a prefix during rule compile.

It’s the closes we can get to pf(4) supporting suffixes natively.

**Comment by marjohn56** [comment]
just looking at the rules.debug...

Just thinking out loud here, if the interface is a tracked interface then should we automatically use/create the alias? We don't need to add an option to the rules creation or do we, if the user also wants to use a ULA as well as, is that likely to cause an issue?

**Comment by fichtner** [comment]
Let’s not overcomplicate this. Prefix merging is also relevant for WAN interface and automating something like this will require a lot of glue that currently does not exist. Adding more code to static rules pages could be a showstopper if we cannot back up the need for it.

**Comment by marjohn56** [comment]
Yes, accept the WAN interface is also relevant, so how do you perceive the alias is created, automatically if it's a dynamic interface or a manual option somewhere?

**Comment by HikariWS** [comment]
What about improving NPTv6? If the problem is concatenating the changing global prefix to LAN devices suffix, then maybe NPTv6 makes it easier as it translates a static LAN prefix (ULA?) to global prefix.

If we can make NPTv6 detect when a global prefix changes, it can update its table and continue translating normally. And we use the static LAN prefix everywhere.

Another suggestion would be to use interface instead of IPv6 address, but of course that doesn't work when we wanna set rules for specific devices addresses.

**Comment by maurice-w** [comment]
Just brainstorming:
- Two new alias types: *Host(s) (tracking)* and *Network(s) (tracking)*.
- These alias types allow you to select the WAN interface they should track.
- *Host(s) (tracking)* look like this: `::1:0:0:0:100` `::2:200:5eff:fe00:53ab`
- *Network(s) (tracking)* look like this: `::2/64` `::40/60`
- When dhcp6c does its thing, we update these aliases by merging them with the delegated prefix.

**Comment by marjohn56** [comment]
Yes, that could work; however found another issue whilst playing. There was another issue reported where the OP said that when the prefix changed the original IA/PD was not removed. I 'tut tutted' that, however he was correct! I'm currently re-mangling dhcp6c to fix that. When I looked at the code there it was, a comment that may have been by Franco ( or someone else  ) that the PD would not get removed as it stood, this left two prefixes on the interface and resulted in RADVD not updating to the correct prefix. Also, we have to look at dhcpdv6, it should not give out leases longer than the lease obtained by dhcp6c...

**Comment by marjohn56** [comment]
@maurice-w - Can you try the attached, I've sent the full source too, but the compiled client is in there. So what am I now seeing that's different.... Downstream client immediately updates its address to the new prefix, probably because radvd is now getting the correct address from the interface, dhcp6c was leaving the old address in place when the prefix was changed.
To achieve this I've forced dhcp6c to restart when an ia or pd change is detected, it just bins of the addresses and starts afresh. To test this in my systems I have my test proc pulling an address prefix from my primary router, one the prefix is obtained and we're into the lease time, then I change the available prefix range on the upstream router, once the renew time comes around the dhcpdv6 response to the renew is 'DH6OPT_STCODE_NOBINDING', which means the existing renew request cannot have that address as it's not available and we force a restart. Now something else appears to play nicer too, in the existing client, should this happen I see it clears the t1/t2 timers but as it cannot get a new lease it retries with the timer lengths extremely short, this results, at least on my system, with it just sending out renews for some time, I think eventually it does pick up a new address but it still does not release the old one, I don't know how long it takes for this to happen because I got bored watching it! :) See if it behaves for you
[dhcp6c.zip](https://github.com/opnsense/core/files/6290967/dhcp6c.zip)

, and if you are able to force a new range try that too.

**Comment by marjohn56** [comment]
Hold hard, I forgot to put back the HUP catch, have done it now, just testing and I'll repost.

**Comment by maurice-w** [comment]
No worries, haven't tested it yet. I'll have to set up some new test VMs first. Don't want to make any dhcp6c experiments with my ISP.

**Comment by marjohn56** [comment]
Fine, I wanted to test it thoroughly here and make sure that nothing unexpected happens. I'm as sure as I can be now that it's behaving as I expect it to. I have also added an ENV var to give the obtained PL and VL lease times, and have written that into interfaces.inc and picked them up in dhcpd.inc, thus the lease times are now pushed down stream to the clients which means, all being good that they renew their leases shortly after the lease has been updated by dhcp6c from upstream. As this is more than just dhcp6c I'll issue a PR for the scripts for testing purposes and attach dhcp6c here.
[dhcp6c.zip](https://github.com/opnsense/core/files/6292866/dhcp6c.zip)

**Comment by marjohn56** [comment]
PR for the scripts: #4916

**Comment by marjohn56** [comment]
And another addition advising the user who has used override what's happening.

![image](https://user-images.githubusercontent.com/17962918/114575219-dabd2b00-9c71-11eb-8c27-ebb4b285d925.png)

**Comment by marjohn56** [comment]
We need to another field to the rules. if the user were to enter a destination address of ::1000, that's fine, but it means squat without the interface that the address is on. So I am proposing that if an address such as ::1000 is entered, i.e. a suffix only then then in the same way as the mask auto expands to give up to a 128 mask a new field appears below the address called 'Destination Interface' which allows you to select the required interface. 

Thoughts please.

**Comment by fichtner** [comment]
To be honest, I explained this clearly. You put the address in the alias, set an interface for the alias to track and use the alias in the rule. And that's it.

**Comment by marjohn56** [comment]
So you are forcing the user to use an Alias then?

**Comment by fichtner** [comment]
Because otherwise we can't support multiple addresses at once (create one per rule instead?) and this way we decouple the address into a table which we can reload in the background without changing the rules.

**Comment by marjohn56** [comment]
Look, I'm not saying we don't use the alias method, I'm looking at all options. If you just want to create a simple rule to open a single port to a single destination it's as quick to enter it in rules as it is to enter an alias and then enter the rule. Frankly I don't really give a ****, I'm just trying to not force a single method.

**Comment by fichtner** [comment]
The alias approach is the one of least resistance all things considered. I am here to manage expectations and avoid future shortcomings.

Supporting a single address seems tempting but given general firewall rule requirements the new feature will be considered incomplete or buggy by anyone not involved here prior.

We were also hoping to avoid adding more code to static PHP pages.


Cheers,
Franco

**Comment by maurice-w** [comment]
I'm with @fichtner on going the alias way (only). Keep it simple.
As for setting an interface for the alias to track: Do you mean a LAN interface or a WAN Interface? I suggest WAN interface. If the alias would track a LAN interface (which is itself tracking a WAN), you wouldn't be able to create aliases for subnets which are not assigned to a local Interface. Which is important for downstream prefix delegation.

**Comment by marjohn56** [comment]
So how is that going to work if you are opening a port to a client on the LAN? You need to prepend the address of the LAN prefix.

**Comment by fichtner** [comment]
The alias setting for the interface of the suffix list is what you merge the address prefix with. It doesn’t matter if it’s WAN or LAN or static or dynamic. It would work either way. It would keep working even when you change the IPv6 type actually.

**Comment by marjohn56** [comment]
What happens when you don't have a WAN prefix then?

**Comment by fichtner** [comment]
You mean what happens when you don’t have an IPv6 address? Not much likely. It could flush the suffix into the table as is to avoid an empty table.

**Comment by marjohn56** [comment]
I'm getting a bit confused here. IF you have mutliple WANs and multiple VLANs you HAVE to be able to specify LAN or WAN interface prefix.

**Comment by marjohn56** [comment]
Therefore in the edit alias you need to be able to select an interface from which to take the prefix.

**Comment by fichtner** [comment]
That‘s what I said a few times now. ;)

**Comment by maurice-w** [comment]
What I'm suggesting is tracking the delegated prefix itself (`/tmp/$_pdinfo`), not the /64 subnet used by a specific interface. Take a closer look at my examples in the comment from 4 days ago. :-)
For _hosts_, you would enter the subnet ID (0..16 bits) + interface ID (64 bits). For _networks_, you would enter a stub prefix + prefix length.
This gets merged with the delegated prefix itself (48..64 bits), not with the subnet used a by a specific LAN interface.

**Comment by marjohn56** [comment]
> That‘s what I said a few times now. ;)

Something is getting lost in translation then! :)

**Comment by marjohn56** [comment]
@maurice-w  - Thats not a problem - but you still need to be able to select the interface. ( head banging against wall repeatedly! )

**Comment by marjohn56** [comment]
Like this.

![image](https://user-images.githubusercontent.com/17962918/114721190-9cd30c00-9d30-11eb-9065-02b4586521b8.png)

**Comment by marjohn56** [comment]
I'm almost there, just cannot get the Interface select to populate with anything...

**Comment by maurice-w** [comment]
@marjohn56 Yes, exactly. 'Interface' would have to be an interface with IPv6 configuration type 'DHCPv6', _not_ 'Track Interface'. Agreed?
Entering ::2000 would be interface ID ::2000 in subnet 0. If you want interface ID ::2000 in e.g. subnet 1, you would enter ::1:0:0:0:2000.

**Comment by fichtner** [comment]
No, basically we just need primary IPv6 of the target interface to merge prefix. Still doesn’t matter how it got this address.

**Comment by marjohn56** [comment]
Get's a bit complex for some too, they'd have to start being able to count it hex.

**Comment by marjohn56** [comment]
@fichtner - So what you see from what I've already done above, is that what you were thinking?

**Comment by maurice-w** [comment]
@fichtner, this would make it impossible to create aliases for subnets not assigned to an interface (downstream prefix delegation).
@marjohn56 Misunderstanding, these are just examples. I'm just saying that what you enter in the 'Content' field is not just the host's 64-bit interface identifier, but anything up to 80 bits (subnet ID + interface ID).

**Comment by marjohn56** [comment]
@maurice-w  -  In that case I would suggest that the suffix has to be entered is in its full form, so a suffix of 80bits is entered as a full address like so:-  ::abab:1234:5678:abcd:ef12 , a 64 bit suffix would be ::1234:5678:abcd:ef12
We take that address, parse it, working out that length and then prefix it with the corresponding length of the prefix needed.

**Comment by maurice-w** [comment]
I think we can just assume anything entered to be `(128 - length($_pdinfo))` in length. If the delegated prefix you get from upstream is e.g. a /56, anything entered for the alias will be treated as 72 bits in length.

That's for host aliases, for network aliases it's a little bit more complicated.

**Comment by marjohn56** [comment]
That's not what @fichtner wants though is it? Look, I'll just carry on with the interface, once I've got that sorted we'll consider the merging.

**Comment by maurice-w** [comment]
If I understand @fichtner's suggestion correctly, I'm not sure how we could implement network aliases like `::40/60`. I might sound like a broken record, but please consider downstream prefix delegation. If OPNsense gets e.g. a /56 delegation from upstream and delegates a /60 to a downstream router, we need to be able to add firewall rules for that /60 network. Otherwise, inbound connections to the downstream router would be impossible.

**Comment by marjohn56** [comment]
No look, I agree with you, I do it all the time testing from my primary down through my test router.

**Comment by marjohn56** [comment]
Erm, I think I've done it. :) @fichtner this is now in the rules.debug, is this what it should look like?

# User Aliases
table <IPv6_Host>  persist  
IPv6_Host = "<IPv6_Host>"
Tst_IPv6_Dynamic = "{ 2a02:8010:6228:f600:0:0:0:2000 }"

**Comment by marjohn56** [comment]
Apparently so, my prefix changed and so has rules debug. I'll tidy it up and PR it for playing with. In its current state it just merges the prefix with the suffix, with a little bit more work @maurice-w we should be able to merge any prefix lengths etc.

**Comment by marjohn56** [comment]
First commit #4919

**Comment by TheRealBecks** [comment]
What's the next step here?

**Comment by TheRealBecks** [comment]
#5284 and this one are stalled :( Or are you waiting for #5574 to be pulled in?

**Comment by fichtner** [comment]
#5284 has been shipped. It's still not perfect but will continue the work in #6158

**Comment by kathampy** [comment]
Can this be enhanced to match a /64 subnet instead of just a single host? I need to create rules for entire subnets on a downstream L3 switch.
